File: android/src/main/java/io/agora/rtc/react/PromiseCallback.kt
package io.agora.rtc.react

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import io.agora.rtc.base.Callback

class PromiseCallback(
  private val promise: Promise?
) : Callback() {
  override fun success(data: Any?) {
    when (data) {
      null, is Boolean, is Int, is String -> promise?.resolve(data)
      is Number -> promise?.resolve(data.toDouble())
      is Array<*> -> promise?.resolve(Arguments.makeNativeArray<Any>(arrayOf(data)))
      is List<*> -> promise?.resolve(Arguments.makeNativeArray(data))
      is Map<*, *> -> {
        val map = mutableMapOf<String, Any?>()
        data.forEach {
          if (it.key is String) {
            map[it.key as String] = it.value
          }
        }
        promise?.resolve(Arguments.makeNativeMap(map))
      }
      else -> promise?.reject(IllegalArgumentException("Could not convert " + data.javaClass))
    }
  }

  override fun failure(code: String, message: String) {
    promise?.reject(code, message)
  }
}

File: ios/RCTAgora/React/PromiseCallback.swift
//
//  PromiseCallback.swift
//  RCTAgora
//
//  Created by LXH on 2020/4/13.
//  Copyright (c) 2020 Syan. All rights reserved.
//

import Foundation
import AgoraRtcKit

@objc(PromiseCallback)
class PromiseCallback: NSObject, Callback {
    private var resolve: RCTPromiseResolveBlock?
    private var reject: RCTPromiseRejectBlock?

    init(_ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        self.resolve = resolve
        self.reject = reject
    }

    func success(_ data: Any?) {
        resolve?(data)
    }

    func failure(_ code: String, _ message: String) {
        reject?(code, message, nil)
    }
}
